gtime/causality/granger_causality.py

```diff
@@ -0,0 +1,213 @@
+            result_df.index = ['F-value', 'p-value', 'degrees of freedom', 'number of shifts']
+            result.append(result_df)
+
+        if 'ssr_chi2' in stat:
+            result_df = pd.DataFrame()
+            fgc2 = len(data_single) * (lr_single_residues - lr_joint_residues) / lr_joint_residues
+            result_df['ssr_chi2test'] = [fgc2, stats.chi2.sf(fgc2, self.max_shift), int(DOF_joint), int(self.max_shift)]
+            result_df.index = ['chi2', 'p-value', 'degrees of freedom', 'number of shifts']
+            result.append(result_df)
+        
+        if 'likelihood_chi2' in stat:
+            result_df = pd.DataFrame()
+            lr_single_loglikelihood = loglikelihood(y_pred=y_pred_single, y_true=data[x].loc[data_single.index])
+            lr_joint_loglikelihood = loglikelihood(y_pred=y_pred_joint, y_true=data[x].loc[data_joint.index])
+            
+            lr = -2 * (lr_single_loglikelihood - lr_joint_loglikelihood)
+
+            result_df['likelihood ratio test'] = [lr, stats.chi2.sf(lr, self.max_shift), int(DOF_joint), int(self.max_shift)]
+            result_df.index = ['chi2', 'p-value', 'degrees of freedom','number of shifts']
+            result.append(result_df)
+
+        if 'zero_F' in stat:
+            result_df = pd.DataFrame()
+            r_matrix = np.column_stack((np.zeros((self.max_shift, self.max_shift)),
+                                        np.eye(self.max_shift, self.max_shift),
+                                        np.zeros((self.max_shift, 1))))
+            
+            y_true = data[x].loc[data_joint.index].values 
+            q_matrix = np.zeros(len(r_matrix)) 
+            
+            params = list(lr_joint.coef_)
+            params.append(lr_joint.intercept_)
+            params = np.array(params) 
+            cparams = np.dot(r_matrix, params[:, None])
+            Rbq = cparams - q_matrix
+            
+            scale = mean_squared_error(y_pred_joint, data[x].loc[data_joint.index])
+            scale = lr_joint_residues/DOF_joint
+            params = params.reshape(1, -1)
+            
+            pinv_wexog = pinv_extended(whiten(data_joint.values.reshape(len(data_joint), -1)))
+            normalized_cov_params = np.dot(pinv_wexog, np.transpose(pinv_wexog)) 
+            
+            cov_p = normalized_cov_params * scale 
+            cov_p = np.dot(r_matrix, np.dot(cov_p, np.transpose(r_matrix)))
+            invcov = np.linalg.pinv(cov_p)
+
+            F = np.dot(np.dot(Rbq.T, invcov), Rbq) 
+            F /= len(r_matrix)
+            F = np.unique(F)[0]
+
+            pvalue = stats.f.sf(F, len(r_matrix), DOF_joint)
+            pvalue = np.unique(pvalue)[0]
+            
+            result_df['F-test'] = [F, pvalue, int(DOF_joint), int(self.max_shift)]
+            result_df.index = ['F-value', 'p-value', 'degrees of freedom', 'number of shifts']
+            result.append(result_df)
+
+        if len(result)==1:
+            return result[0]
```

Comment:
The `fit()` method should return `self`

(Get some inspiration from https://github.com/scikit-learn/scikit-learn/blob/b194674c4/sklearn/covariance/_shrunk_covariance.py#L63)